engine/src/caching_client.rs

```diff
@@ -0,0 +1,224 @@
+					let mut in_flight: HashMap<
+						Key,
+						Vec<oneshot::Sender<Result<ResultType, anyhow::Error>>>,
+					> = HashMap::default();
+					let (cache_sender, mut cache_receiver) =
+						mpsc::unbounded_channel::<(Key, Result<ResultType, anyhow::Error>)>();
+					loop {
+						tokio::select! {
+							biased;
+							Some(()) = cache_invalidation_receiver.recv() => {
+								// We only clear the cache, we don't want to cancel in_flight requests, this way we are sure that every request will eventually complete
+								// regardless of the time it takes to complete
+								cache.clear();
+							},
+							Some((key, value)) = cache_receiver.recv() => {
+								if let Ok(ref value) = value {
+									cache.insert(key.clone(), value.clone());
+								}
+								if let Some(senders) = in_flight.remove(&key) {
+									for sender in senders {
+										let to_send = match &value {
+											Ok(val) => Ok(val.clone()),
+											Err(e) => Err(anyhow::anyhow!(e.to_string())),
+										};
+										let _ = sender.send(to_send);
+									}
+								}
+							},
+							Some((future, request_key, sender)) = request_receiver.recv() => {
+								if let Some(value) = cache.get(&request_key)  {
+										let _ = sender.send(Ok(value.clone()));
+								} else if let Some(result_senders) = in_flight.get_mut(&request_key) {
+									result_senders.push(sender);
+								} else {
+									let sender_result = cache_sender.clone();
+									in_flight.insert(request_key.clone(), vec![sender]);
+									scope.spawn(async move {
+										let result = future.await;
+										let _ = sender_result.send((request_key, result));
+										Ok(())
+									})
+								}
+
+							},
+						}
+					}
+				}
+				.boxed()
+			})
+		});
+
+		(Self { request_sender, client }, cache_invalidation_sender)
+	}
+
+	pub(crate) async fn get(
+		&self,
+		future: TypedFutureGenerator<ResultType, Client>,
+		key: Key,
+	) -> Result<ResultType, anyhow::Error> {
+		let (tx, rx) = oneshot::channel::<Result<ResultType, anyhow::Error>>();
+		let client = self.client.clone();
+		let future = future(client);
+		self.request_sender.send((future, key, tx)).unwrap();
+		let result = timeout(MAX_DURATION, rx).await???;
+		Ok(result)
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use crate::caching_client::CachingRequest;
+	use anyhow::Error;
+	use cf_utilities::task_scope::{task_scope, Scope};
+	use futures_util::FutureExt;
+	use rand::random;
+	use std::{
+		sync::atomic::{AtomicU32, Ordering},
+		time::Duration,
+	};
+	use tokio::time::sleep;
+
+	static A: AtomicU32 = AtomicU32::new(0);
+
+	trait Rpc {
+		async fn a(&self) -> anyhow::Result<u32>;
+	}
+	#[derive(Clone, Default)]
+	struct Client {}
+	impl Rpc for Client {
+		async fn a(&self) -> anyhow::Result<u32> {
+			sleep(Duration::new(1, 0)).await;
+			A.fetch_add(1, Ordering::Relaxed);
+			Ok(random::<u32>())
+		}
+	}
+
+	#[tokio::test]
+	async fn internal_client_called_once() {
```

Comment:
A bit harder to test, but I think we should ensure that we clearly test 2 cases of:
1. Request for something already cached only results in 1 request (pulled from cache)
2. Request for something already *in flight* results in 1 request (gets a clone of receiver)


Overall, they are nice tests btw 👌